controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -395,6 +396,44 @@ func (c *Cluster) ReconcileKubeletRBACRole(ctx context.Context, version semver.V
 	return nil
 }
 
+// ClusterStatus holds stats information about the cluster.
+type ClusterStatus struct {
+	// Nodes are a total count of nodes
+	Nodes int32
+	// ReadyNodes are the count of nodes that are reporting ready
+	ReadyNodes int32
+	// HasKubeadmConfig will be true if the kubeadm config map has been uploaded, false otherwise.
+	HasKubeadmConfig bool
+}
+
+// ClusterStatus returns the status of the cluster.
+func (c *Cluster) ClusterStatus(ctx context.Context) (ClusterStatus, error) {
+	status := ClusterStatus{}
+
+	// count the control plane nodes
+	nodes, err := c.getControlPlaneNodes(ctx)
+	if err != nil {
+		return status, err
+	}
+
+	for _, node := range nodes.Items {
+		nodeCopy := node
+		status.Nodes++
+		if util.IsNodeReady(&nodeCopy) {
+			status.ReadyNodes++
+		}
+	}
+
+	// find the kubeadm conifg
+	kubeadmConfigKey := ctrlclient.ObjectKey{
+		Name:      "kubeadm-config",
+		Namespace: metav1.NamespaceSystem,
+	}
+	err = c.Client.Get(ctx, kubeadmConfigKey, &corev1.ConfigMap{})
+	status.HasKubeadmConfig = err == nil
```

Comment:
Should we return an error if the `err` is anything other that IsNotFound?

Reply:
An interesting question. I think this is a matter of what semantics this should have.

I could introduce a third unknown state and set HasKubeadmConfig to that if the error is anything other than IsNotFound and set HasKubeadmConfig = false if the error IsNotFound and true otherwise. Or this could return some other error (anything besides IsNotFound) and then we'd have to make an assumption:

Do we update HasKubeadmConfig? If we don't, the value could be stale If we do, the value could be wrong. 

I'm leaning towards adding a third state.

What do you think?

Reply:
yeah what's there is fine, it just stood out because we're eating any error effectively, which could be misleading, I'm fine adding a TODO for now and tackle it later